qgrid/grid.py

```diff
@@ -81,3 +85,108 @@ def _ipython_display_(self):
             display_javascript(raw_js, raw=True)
         except Exception as err:
             display_html('ERROR: {}'.format(str(err)), raw=True)
+
+
+class QGridWidget(widgets.DOMWidget):
+    _view_name = Unicode('QGridView', sync=True)
+    _df_json = Unicode('', sync=True)
+    _column_types_json = Unicode('', sync=True)
+    _loop_guard = Bool(False)
+    _index_name = Unicode('')
+    _cdn_base_url = Unicode("/nbextensions/qgridjs", sync=True)
+
+    df = Instance(pd.DataFrame)
+    editable = Bool(True, sync=True)
+    remote_js = Bool(False)
+
+    def _df_changed(self):
+        """Build the Data Table for the DataFrame."""
+        if self._loop_guard:
+            return
+        df = self.df.copy()
+
+        # register a callback for custom messages
+        self.on_msg(self._handle_qgrid_msg)
+
+        if not df.index.name:
+            df.index.name = 'Index'
+
+        if type(df.index) == pd.core.index.MultiIndex:
+            df.reset_index(inplace=True)
+        else:
+            df.insert(0, df.index.name, df.index)
+
+        self._index_name = df.index.name
+
+        tc = dict(np.typecodes)
+        for key in np.typecodes.keys():
+            if "All" in key:
+                del tc[key]
+
+        column_types = []
+        for col_name, dtype in df.dtypes.iteritems():
+            column_type = {'field': col_name}
+            for type_name, type_codes in tc.items():
+                if dtype.kind in type_codes:
+                    column_type['type'] = type_name
+                    break
+            column_types.append(column_type)
+        self._column_types_json = json.dumps(column_types)
+
+        precision = pd.get_option('display.precision') - 1
+
+        self._df_json = df.to_json(
+                orient='records',
+                date_format='iso',
+                double_precision=precision,
+            )
+
+    def _remote_js_changed(self):
+        if self.remote_js:
+            self._cdn_base_url = REMOTE_URL
+        else:
+            self._cdn_base_url = "/nbextensions/qgridjs"
+
+    def add_row(self, value=None):
+        """Append a row at the end of the dataframe."""
+        df = self.df
+        if not df.index.is_integer():
+            msg = 'alert("Cannot add a row a table with a non-integer index")'
+            display(Javascript(msg))
+            return
+        last = df.loc[df.index[-1], :]
```

Comment:
Ha!, perhaps, I confess I do not use pandas regularly, so any help in that regard is appreciated.